### PR TITLE
Safari mobile audiocontext suspended fix

### LINKED ISSUE #338: iPhone: unplugging/replugging headphones causes audio problems
While in a space with another user, user A unplugs headphones (standard issue Apple wired earbuds) then plugs them back in. The result is that they now cannot hear user B. It seems that sometimes it causes garbled audio instead of no audio. User B can still hear user A just fine as it doesnt seem to affect user A’s headphone mic—(or perhaps it’s using the built-in iPhone mic at this point?)
This persists until user B refreshed their page, which seems strange since the assumption is its an issue on user A’s end.
Then, to make things more interesting, user C joined and subsequently plugged in headphones, causing user B to not hear user C while user A could hear user C just fine.
The bottom line:
plugging in headphones post-join is likely to cause audio issues.
This issue was reproducible on two users’ iPhones. (6s)

### PR DESCRIPTION
**Pull request by keianhzo**
Fixes #338 Hack for a Safari Mobile issue where unplugging/plugging headphones may leave the Audio context in a suspended state.

Based on this contribution from @vincentfretin: https://github.com/mozilla/hubs/issues/338#issuecomment-765356433

I couldn't test this on a physical device so iOS Safari testing is needed.

### CODE REVIEW

**Comment by keianhzo** [comment]
@vincentfretin This PR is based on a contribution of yours in case you want to take a look, review or test.

**Review by vincentfretin — COMMENTED** [review]
I don't think you need to check for browser here. You should always add the this.audioContext.onstatechange listener. It's harmless for other browsers.

File: src/systems/audio-system.js [hunk 1]
```diff
@@ -154,6 +144,11 @@ export class AudioSystem {
     sourceNode.connect(gainNode);
     gainNode.connect(this.outboundGainNode);
     this.audioNodes.set(id, { sourceNode, gainNode });
+
+    document.getElementById("avatar-rig").messageDispatch.receive({
+      type: "log",
+      messageType: LogMessageType.audioResumed
+    });
```
[the remark this answers is not in the document]
Comment (keianhzo) [reply src/systems/audio-system.js:151]: Oops, that was just a test.

File: src/systems/audio-system.js [hunk 2]
```diff
@@ -120,28 +123,15 @@ export class AudioSystem {
...
-            enableChromeAEC(sceneEl.audioListener.gain);
-          }
-
-          document.body.removeEventListener("touchend", resume, false);
-          document.body.removeEventListener("mouseup", resume, false);
-        }
-      }, 0);
-    };
+    // Safari Mobile fix
+    if (AFRAME.utils.device.isMobile() && AFRAME.utils.device.isIOS()) {
```
[the remark this answers is not in the document]
Comment (keianhzo) [reply src/systems/audio-system.js]: Updated.

**Comment by keianhzo** [comment]
@vincentfretin I have removed the iOS specific check. Still can't test so if anybody can please do and report.

**Comment by vincentfretin** [comment]
Yes, better. I'll try to test that soon with #3944

File: src/systems/audio-system.js [hunk 3]
```diff
@@ -164,4 +152,49 @@ export class AudioSystem {
...
+    this.audioContext.onstatechange = () => {
+      console.log(`AudioContext state changed to ${this.audioContext.state}`);
+      if (this.audioContext.state === "suspended") {
+        // When you unplug the headphone or when the bluetooth headset disconnects on
+        // iOS Safari or Chrome, the state changes to suspended.
+        // Chrome Android doesn't go in suspended state for this case.
+        document.getElementById("avatar-rig").messageDispatch.receive({
+          type: "log",
+          messageType: LogMessageType.audioSuspended
+        });
```
Comment (brianpeiris) [inline src/systems/audio-system.js:187]: nit: Looks like MessageDispatch has a `log` method that you could use to abstract some of this away.

**Review by brianpeiris — APPROVED** [review]
Only one minor comment. LGTM otherwise.

File: src/systems/audio-system.js (hunk 3 above)
Comment (keianhzo) [reply src/systems/audio-system.js]: True, fixed. Thanks!

**Comment by vincentfretin** [comment]
Test on iPad with iOS 14.3

Scenario:
- insert jack on iPad, all is working fine
- unplug jack

With master:
Results: mic volume indicator doesn't work, we don't hear others (the head doesn't scale)

With this PR:
Results:
- mic volume indicator doesn't work, we don't hear others (the head doesn't scale)
- "Audio has been resumed" log shows up (instead of "Audio has been suspended, click somewhere in the room to resume the audio.", weird)
- I click and move in the scene
- "Audio has been resumed" log shows up, the mic volume works, I can hear others.

The changes works but with the wrong first message, weird.

**Comment by vincentfretin** [comment]
I don't know what going on with `LogMessageType.audioSuspended` why it doesn't show the correct message
If I replace
`document.getElementById("avatar-rig").messageDispatch.log(LogMessageType.audioSuspended)`
by for example
`document.getElementById("avatar-rig").messageDispatch.log(LogMessageType.audioNormalizationNaN)`
I see the "audioNormalization command needs a valid number parameter." message, so good.

**Comment by vincentfretin** [comment]
If I add a new message in `const logMessages = defineMessages` in `ChatSidebar.js`
I actually see this new message for both `LogMessageType.audioSuspended` and `LogMessageType.audioResumed`
so it seems that `document.getElementById("avatar-rig").messageDispatch.log` doesn't take care of the argument and just use the last message in the `logMessages` list.

**Comment by vincentfretin** [comment]
I quickly looked at `src/message-dispatch.js` I don't understand how the translation machinery works. So I'll let you look into this.

**Comment by keianhzo** [comment]
Thanks for the review @vincentfretin! Good catch, I was missing adding the messages to the `LogMessageType` object. I've added that and now it should be working fine.

**Comment by vincentfretin** [comment]
Yes! This is good now.
IMG_0083
You can merge.

**Comment by keianhzo** [comment]
Great, I'll merge then. Thanks a lot for your support!

**Pull request merged**